Add NativeAOT on Android test

jtschuster opened the pull request:
Adds a project to run a NativeAOT shared library on Android.

The android test builds a single native library that exports the methods that are expected by the monodroid tests (InitRuntime, SetEnv, ExecEntryPoint, FreeUnmanagedResources). JNI struct definitions are copied to get arg strings from the JNIEnv object.

ApkBuilder no longer needs to compile the monodroid native library, but it does need to add the java .jar from the cryptography library to the apk bundle in order to run crypto operations.

Build targets were updated to account for the new NativeAOT tests, and a new job was added for NativeAOT on android tests.

jtschuster committed:
NativeAOT Android test working locally

jtschuster committed:
Undo extra changes, add reference in tests.proj

jtschuster committed:
Remove emtpy 'if' case

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into AndroidAppBuilderNaot

jkotas commented on src/coreclr/tools/Common/CommandLineHelpers.cs:
```diff
@@ -82,6 +82,7 @@ public static TargetOS GetTargetOS(string token)
                 "ios" => TargetOS.iOS,
                 "tvossimulator" => TargetOS.tvOSSimulator,
                 "tvos" => TargetOS.tvOS,
+                "android" => TargetOS.Linux,
```
Is this needed given that we are replacing Android with Linux here? https://github.com/dotnet/runtime/blob/1e860a4bd3915f35d51cc079cff212322a6991e7/src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets#L237-L238

We should do this substitution in a single place. Or if we believe that ilc needs to know about Android, we should go all the way and add it to TargetOS enum.

jkotas commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.SingleEntry.targets:
```diff
@@ -31,7 +31,8 @@
     <_linuxLibcFlavor Condition="$(_targetOS.StartsWith($(_linuxToken)))">$(_targetOS.SubString($(_linuxToken.Length)))</_linuxLibcFlavor>
-    <_targetOS Condition="$(_targetOS.StartsWith($(_linuxToken)))">linux</_targetOS>
+    <_linuxLibcFlavor Condition="'$(_targetOS)' == 'android'">bionic</_linuxLibcFlavor>
+    <_targetOS Condition="$(_targetOS.StartsWith($(_linuxToken))) or $(_targetOS) == 'android'">linux</_targetOS>
```
Another redundant (?) Android -> substitution

jtschuster committed:
Remove extra android->linux translations

davidnguyen-tech commented on src/tasks/AndroidAppBuilder/ApkBuilder.cs:
```diff
@@ -263,7 +266,7 @@ public ApkBuilder(TaskLoggingHelper logger)
             }
 
-            if (!File.Exists(runtimeLib))
+            if (!File.Exists(runtimeLib) && !IsNativeAOT)
```
Is this `!IsNativeAOT` check necessary, given that it's already evaluated at line 249? It might be redundant.

jtschuster replied:
This can be removed.

jtschuster committed:
Remove redundant IsNativeAOT check

jtschuster committed:
Fix build to work in CI

- Don't use BundleDir for android bundle dir: This is PublishDir when TestSingleFile=true (which it is). Android deletes this dir before packaging, deleting the files that were just published.

- Don't add the singleFileTestRunner.cs when the test sets IsFunctionalTest=true

- Don't run regular nativeaot smoke tests on android yet.

- Enable android tests to publish as libs when TestSingleFile=true (default when TestNativeAOT=true)

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into AndroidAppBuilderNaot

jtschuster committed:
Fix apkbuilder.cs errors after merge

jtschuster committed:
Undo deleted line in ILC.targets, build libs.tests in pipeline

jtschuster committed:
Add missing args to build command

jtschuster committed:
Set CppCompilerAndLinker to NDK clang when targeting android

jtschuster committed:
Ensure Android build runs after CopyNativeBinary

jtschuster committed:
Merge branch 'main' into AndroidAppBuilderNaot

risc-vv commented:
@dotnet/samsung Could you please take a look? These changes may be related to riscv64.

kotlarmilos commented on eng/pipelines/runtime.yml:
```diff
@@ -993,6 +993,48 @@ extends:
+      # Android arm64 devices and x64 emulators
+      # Build the whole product using CoreCLR and run functional tests
+      #
+      - template: /eng/pipelines/common/platform-matrix.yml
```
Consider moving this to runtime-extra-platforms pipeline to reduce the load on devices and emulators.

kotlarmilos commented on eng/testing/tests.singlefile.targets:
```diff
@@ -1,6 +1,6 @@
 <Project>
   <PropertyGroup>
-    <OutputType>Exe</OutputType>
+    <OutputType Condition="'$(TargetOS)' != 'android'">Exe</OutputType>
```
Why is this needed?

kotlarmilos commented on eng/targetingpacks.targets:
```diff
@@ -66,14 +66,14 @@
                       LatestRuntimeFrameworkVersion="$(ProductVersion)"
                       RuntimePackNamePatterns="$(LocalFrameworkOverrideName).Runtime.NativeAOT.**RID**"
-                      RuntimePackRuntimeIdentifiers="ios-arm64;iossimulator-arm64;iossimulator-x64;tvos-arm64;tvossimulator-arm64;tvossimulator-x64;maccatalyst-arm64;maccatalyst-x64;linux-bionic-arm64;linux-bionic-x64;osx-arm64;osx-x64"
+                      RuntimePackRuntimeIdentifiers="ios-arm64;iossimulator-arm64;iossimulator-x64;tvos-arm64;tvossimulator-arm64;tvossimulator-x64;maccatalyst-arm64;maccatalyst-x64;linux-bionic-arm64;linux-bionic-x64;android-x64;android-arm64;osx-arm64;osx-x64"
```
Do the tests consume runtime packs or local artifacts? If they use local artifacts, I don't think these changes are necessary.

kotlarmilos commented on src/libraries/tests.proj:
```diff
@@ -607,6 +607,10 @@
+    <SmokeTestProject Include="$(RepoRoot)src\tests\FunctionalTests\Android\Device_Emulator\NativeAOT\*.Test.csproj"/>
+  </ItemGroup>
+
   <ItemGroup Condition="('$(TargetOS)' == 'iossimulator' or '$(TargetOS)' == 'tvossimulator') and '$(RuntimeFlavor)' == 'CoreCLR'">
```
I think this is a merge conflict and shouldn't be needed.

jtschuster replied:
What part? 610-612 is needed to build and test the new nativeaot test project

jtschuster committed:
Test CI in Debug config

jtschuster committed:
Merge branch 'main' into AndroidAppBuilderNaot

jtschuster committed:
Move leg to runtime-extra-platforms

jtschuster committed:
Merge branch 'AndroidAppBuilderNaot' of https://github.com/jtschuster/runtime into AndroidAppBuilderNaot

jtschuster committed:
Remove jobs from runtime

jtschuster commented:
/azp run runtime-extra-platforms

MichalStrehovsky commented:
Is the issue you're seeing on Release this one?

```
java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "_ZN6Thread20DeferTransitionFrameEv" referenced by "/data/app/net.dot.Android.Device_Emulator.NativeAOT.Test-X0GLQAbvFe_cu9kFvBxXiQ==/lib/x86_64/libAndroid.Device_Emulator.NativeAOT.Test.so"...
```

It might be fixable by adding `#include "thread.inl"` to interloplibinterface_java.cpp

jtschuster committed:
Exclude nativeaot test on non-nativeaot builds

jtschuster commented:
/azp run runtime-extra-platforms

jtschuster committed:
Add thread.inl to interoplibinterface_java.cpp to fix loadLibrary failure

jtschuster committed:
Remove android rids from targetpack rids

jtschuster committed:
Run release build in CI

jtschuster committed:
Use AppBundle as AndroidBundleDir as Helix expects

jtschuster committed:
Merge branch 'main' into AndroidAppBuilderNaot

jtschuster commented:
/azp run runtime-extra-platforms

jtschuster committed:
Tighten OutputType condition in singlefile test targets

filipnavara commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -47,7 +47,7 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-bionic-')) or $(CrossCompileRid.StartsWith('android'))">android21</CrossCompileAbi>
       <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-musl-')) or $(CrossCompileRid.StartsWith('alpine-'))">musl</CrossCompileAbi>
       <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-arm'">gnueabihf</CrossCompileAbi>
       <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-bionic-arm'">androideabi21</CrossCompileAbi>
```
```suggestion
      <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-bionic-arm' or '$(CrossCompileRid)' == 'android-arm'">androideabi21</CrossCompileAbi>
```

steveisok commented on eng/pipelines/extra-platforms/runtime-extra-platforms-android.yml:
```diff
@@ -114,3 +114,45 @@ jobs:
+# Android arm64 devices and x64 emulators
+# Build the whole product using CoreCLR and run functional tests
+#
+- template: /eng/pipelines/common/platform-matrix.yml
```
Consider adding to eng/pipelines/runtime-androidemulator.yml and runtime-android.yml as well. These pipelines are for when you only want to run against android and not the entire extra-platforms kitchen sink.

jtschuster replied:
Added to runtime-extra-platforms-androidemulator.yml. Looks like runtime-androidemulator.yml and runtime-android.yml pull in runtime-extra-platforms-androidemulator.yml and runtime-extra-platforms-android.yml, respectively.

steveisok replied:
Yep, and you can also trigger them via `/azp run`

steveisok commented on src/tests/FunctionalTests/Android/Device_Emulator/NativeAOT/Android.Device_Emulator.NativeAOT.Test.csproj:
```diff
@@ -0,0 +1,24 @@
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+  </PropertyGroup>
+
+  <ItemGroup>
```
Put these in a common place?

jtschuster committed:
Add job to runtime-extra-platforms-androidemulator.yml

jtschuster committed:
Update CrossCompileAbi condition

jtschuster committed:
Merge branch 'AndroidAppBuilderNaot' of https://github.com/jtschuster/runtime into AndroidAppBuilderNaot

filipnavara commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -47,10 +47,10 @@ The .NET Foundation licenses this file to you under the MIT license.
       <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-musl-')) or $(CrossCompileRid.StartsWith('alpine-'))">musl</CrossCompileAbi>
       <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-arm'">gnueabihf</CrossCompileAbi>
-      <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-bionic-arm'">androideabi21</CrossCompileAbi>
+      <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-bionic-arm' or '$(CrossCompileRid)' == 'android-arm'>">androideabi21</CrossCompileAbi>
```
```suggestion
      <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-bionic-arm' or '$(CrossCompileRid)' == 'android-arm'">androideabi21</CrossCompileAbi>
```

jtschuster committed:
Update src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets

Co-authored-by: Filip Navara <filip.navara@gmail.com>

jtschuster commented:
/azp run runtime-androidemulator

jtschuster committed:
Use UseNativeAOTRuntime property to test if using nativeaot instead of creating a nativeaot RuntimeFlavor

jtschuster committed:
DirectPInvoke and DirectPInvokeList is not plural in naot

jtschuster committed:
Merge branch 'AndroidAppBuilderNaot' of https://github.com/jtschuster/runtime into AndroidAppBuilderNaot

jtschuster commented:
/azp run runtime-androidemulator

jtschuster committed:
Add null checks in JNI code

jtschuster committed:
Condition CoreCLR tests on the build not being for NativeAOT

jtschuster committed:
Run android tests unconditionally in extra-platforms builds

jtschuster commented on src/tasks/AndroidAppBuilder/Templates/monodroid-nativeaot.cs:
```diff
@@ -0,0 +1,422 @@
+            string result = Marshal.PtrToStringUTF8((nint)chars);
+            if (isCopy != 0)
+                NativeInterface->ReleaseStringUTFChars(thisptr, str, chars);
+            return result;
```
Made the return type nullable.

jtschuster commented:
/azp run runtime-androidemulator

jtschuster committed:
Copy Java files to publish dir, don't search for them in build dir

filipnavara reviewed (approved):
I tested that it makes the Android runtime pack into functional state that can be consumed on the workload side and used to build a whole app. Obviously there's still work that has to be done on the consumption side (dotnet/android) but having an end-to-end scenrio running is a good start:

<img width="108" height="240" alt="image" src="https://github.com/user-attachments/assets/e5c84f2c-3365-4cb7-b2a7-14efa9a1247a" />


One thing that still needs to be resolved is who is responsible for exporting the JNI symbol for `Java_net_dot_android_crypto_DotnetProxyTrustManager_verifyRemoteCertificate` in case of the builds that link the crypto libraries statically. Presumably that can and will be done on the Android workload side with something like `<IlcArg Include="--export-dynamic-symbol:Java_net_dot_android_crypto_DotnetProxyTrustManager_verifyRemoteCertificate" />`.

filipnavara commented:
> @filipnavara maybe this will solve the JNI export problem for crypto: #108513

Yes, that's part of the solution. There's an issue for CoreCLR integration filed at https://github.com/dotnet/android/issues/10324 that has a bit more context.

jtschuster commented:
/azp run runtime-androidemulator

filipnavara commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -47,10 +47,10 @@ The .NET Foundation licenses this file to you under the MIT license.
+      <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-bionic-')) or $(CrossCompileRid.StartsWith('android-'))">android21</CrossCompileAbi>
       <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-musl-')) or $(CrossCompileRid.StartsWith('alpine-'))">musl</CrossCompileAbi>
       <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-arm'">gnueabihf</CrossCompileAbi>
-      <CrossCompileAbi Condition="'$(CrossCompileRid)' == 'linux-bionic-arm'">androideabi21</CrossCompileAbi>
```
We need the same adjustment for bionic and android. That's what the change does (even if we don't produce android-arm packs today we may do so in future).

agocke replied:
Doesn’t this remove the change in abi for bionic? Or is that fixed somewhere else?

filipnavara replied:
It doesn't. It's literally on the next line.

filipnavara commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets:
```diff
@@ -47,10 +47,10 @@ The .NET Foundation licenses this file to you under the MIT license.
       <CrossCompileAbi>gnu</CrossCompileAbi>
-      <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-bionic-'))">android21</CrossCompileAbi>
+      <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-bionic-')) or $(CrossCompileRid.StartsWith('android-'))">android21</CrossCompileAbi>
       <CrossCompileAbi Condition="$(CrossCompileRid.StartsWith('linux-musl-')) or $(CrossCompileRid.StartsWith('alpine-'))">musl</CrossCompileAbi>
```
Long-term we probably don't need both, but historically `linux-bionic` existed for MonoVM and NativeAOT. In case of MonoVM it was used for CI testing. In case of NativeAOT it was an actual shipped product. It's too late in the release cycle to remove it. Probably all the use cases for `linux-bionic` can be serviced with `net10.0` TFM with `android-*` RID (as opposed to `net10.0-android` TFM which brings in the workload logic).

am11 replied:
Yup. We discussed it earlier in the release cycle https://github.com/dotnet/runtime/issues/106748#issuecomment-2748294361 and there was some hope; probably there still is. 😅

We are maintaining linux-bionic (basically lowfat android variant) across multiple repos as if this platform is important, only to find out nobody uses (or should be using) this RID in real-world android apps.

jtschuster replied:
The alpine- rid can be removed in a different change.

It seems like the linux-bionic vs android RID issue requires some more discussion, so I'll leave it as is for now.
